jina/proto/message.py

```diff
@@ -0,0 +1,364 @@
+        elif isinstance(envelope, jina_pb2.Envelope):
+            self.envelope = envelope
+        else:
+            # otherwise delay it to after request is built
+            self.envelope = None
+
+        if isinstance(request, bytes):
+            self.request = LazyRequest(request, self.envelope)
+            self._size += sys.getsizeof(request)
+        elif isinstance(request, jina_pb2.Request):
+            self.request = request  # type: Union['LazyRequest', 'jina_pb2.Request']
+        else:
+            raise TypeError(f'expecting request to be bytes or jina_pb2.Request, but receiving {type(request)}')
+
+        if envelope is None:
+            self.envelope = self._add_envelope(*args, **kwargs)
+            # delayed assignment, now binding envelope to request
+            if isinstance(self.request, LazyRequest):
+                self.request._envelope = self.envelope
+
+        if self.envelope.check_version:
+            self._check_version()
+
+    def as_pb_object(self) -> 'jina_pb2.Message':
+        r = jina_pb2.Message()
+        r.envelope.CopyFrom(self.envelope)
+        if isinstance(self.request, jina_pb2.Request):
+            req = self.request
+        else:
+            req = self.request.as_pb_object()
+        r.request.CopyFrom(req)
+        return r
+
+    @property
+    def is_data_request(self) -> bool:
+        """check if the request is not a control request
+
+        .. warning::
+            If ``request`` change the type, e.g. by leveraging the feature of ``oneof``, this
+            property wont be updated. This is not considered as a good practice.
+        """
+        return self.envelope.request_type != 'ControlRequest'
+
+    def _add_envelope(self, pod_name, identity, num_part=1, check_version=False,
+                      request_id: str = None, request_type: str = None,
+                      compress: str = 'NONE', compress_hwm: int = 0, compress_lwm: float = 1., *args,
+                      **kwargs) -> 'jina_pb2.Envelope':
+        """Add envelope to a request and make it as a complete message, which can be transmitted between pods.
+
+        .. note::
+            this method should only be called at the gateway before the first pod of flow, not inside the flow.
+
+
+        :param pod_name: the name of the current pod
+        :param identity: the identity of the current pod
+        :param num_part: the total parts of the message, 0 and 1 means single part
+        :param check_version: turn on check_version 
+        :return: the resulted protobuf message
+        """
+        envelope = jina_pb2.Envelope()
+        envelope.receiver_id = identity
+        if isinstance(self.request, jina_pb2.Request) or (request_id and request_type):
+            # not lazy request, so we can directly access its request_id without worrying about
+            # triggering the deserialization
+            envelope.request_id = request_id or self.request.request_id
```

Comment:
being it a `private` function, and only used in one place, why is there the possibility that `request_id` is different than `self.request.request_id`?

Reply:
remember protobuf do not init any default value and there is no way to force them equal. 
also note that, envelope do not exist in client side, on client side, only `request.request_id` is available